Userland virtual alarm

alevy opened the pull request:
Fixes #290 

This PR adds support for multiple concurrent alarms (not timers, but we can implement timers on top of this) in userland. It does so with a small extention to the Timer system call driver and a userland library that uses a binary heap to keep track of outstanding alarms.

In order, the commits do the followin:

  1. Add commands to the Timer system-call driver. One that returns the frequency of the clock and another that sets an alarm for an absolute tic value.

  2. An initial implementation of a userland virtual alarm on top of those calls. It turns out it be much trickier (slash maybe infeasible) to virtualize a straight up timer rather than an alarm (we already kind of knew this). It's nearly complete except it doesn't account for clock wrapping yet (just some annoying math to get right and also only happens once every 8 years on the SAM4L). There are also some inefficiencies in how the heap is implemented (plus probably some bugs if I'm being honest).

  3. An example app that blinks two leds at different intervals concurrently.

## You're assistance is requested!

What should the interface look like? What about the division between `virtual_timer.h` and `timer.h`? E.g. should `delay_ms` go in `virtual_timer` (and _be_ a virtual timer) and leave `timer.h` to only include low level calls?

I suspect we will move some or all of the virtualization code into the kernel at some point in the future, but it's probably worth getting the interface right now.

bradjc commented:
My initial thought is that applications shouldn't have to think about `virtual_timer.h` vs `timer.h`. I would expect there to only be `timer.h` and for its high-level functions to be virtualized.

adkinsjd commented:
I agree with Brad. Additionally won't using timer.h at all break any virtual timers a user has running? As a user I think that would be unexpected.

alevy commented:
OK, refactored significantly. I moved things to an `alarm` library and `internal/alarm` library. The non-internal one is virtualized alarms and includes `delay_ms`. The internal ones are basically direct system calls. Still needs some comments and testing (help with testing please?).

The interface basically looks like this now:

```c
// invoke callback at a specific clock value
alarm_t *alarm_start(uint32_t expiration, subscribe_cb, void*);

// invoke callback in `ms` milliseconds
alarm_t *alarm_in(uint32_t ms, subscribe_cb, void*);

// invoke callback repeatedly, every `ms` milliseconds
alarm_repeating_t* alarm_every(uint32_t ms, subscribe_cb, void*);

// delay_ms basically wrappes `alarm_in` as a synchronous call
void delay_ms(uint32_t ms);
```

bradjc commented on userland/examples/tests/hail/main.c:
```diff
@@ -9,14 +9,19 @@
 #include <nrf51_serialization.h>
 #include <si7021.h>
-#include <timer.h>
+#include <si7021.h>
```
Lot of duplicated includes here.

bradjc commented on userland/examples/tests/fxos8700cq/main.c:
```diff
@@ -1,13 +1,16 @@
+#include <stdbool.h>
 #include <stdbool.h>
 #include <stdint.h>
+#include <stdint.h>
```
Duplicated includes.

bradjc commented on userland/examples/sensors/main.c:
```diff
@@ -1,13 +1,14 @@
-#include <timer.h>
 #include <tmp006.h>
 #include <tock.h>
+#include <tock.h>
```
Duplicated include.

bradjc commented on userland/examples/rf233/main.c:
```diff
@@ -4,11 +4,12 @@
+#include <alarm.h>
 #include <gpio.h>
 #include <led.h>
+#include <led.h>
```
Duplicated include.

bradjc commented on userland/examples/multi-timer-blink/main.c:
```diff
@@ -0,0 +1,22 @@
+
+  alarm_every(slow_interval, event_cb, (void*)0);
+  alarm_every(fast_interval, event_cb, (void*)1);
+}
```
Thoughts on moving this to `tests` as something like `multi_timer_test`? Might be more useful to have a general test app rather than two blink apps.

alevy replied:
Yes makes sense

alevy commented on userland/libtock/alarm.h:
```diff
@@ -0,0 +1,32 @@
+
+typedef struct alarm alarm_t;
+
+alarm_t *alarm_start(uint32_t expiration, subscribe_cb, void*);
```
Agree that alarm_at it's better. And should move the `alarm_read` function to this module (from the internal one)

alevy commented:
Comments coming soon

alevy commented:
@bradjc ready for more reviewing after fixing some bugs and adding comments in `alarm.h`.

alevy commented:
@adkinsjd and @ppannuto that you haven't yet reviewed/tested this tells me you may have oversold how useful this would be for you :stuck_out_tongue_closed_eyes: I don't even know which signpost app would benefit from this to make sure it works well with it.

bradjc commented:
I can't build the `multi_alarm_test` app. Travis should have errored.

phil-levis commented:
I'm also confused by the alarm/timer name switch. What is the difference? When I look at the Rust traits, an Alarm is a single event, in terms of a fixed absolute counter value. Timers, in contrast, allow specifying time relative to current time and having repeating events. I.e., an alarm is a thin veneer on top of a compare interrupt while a timer does things like handle overflow, repeat, etc.

But the alarm.h interface has absolute, relative, and repeating. Why is it an alarm?

ppannuto commented on capsules/src/timer.rs:
```diff
@@ -91,54 +112,89 @@ impl<'a, A: Alarm> Driver for TimerDriver<'a, A> {
+                            let num_armed = self.num_armed.get();
+                            self.num_armed.set(num_armed - 1);
+                            if num_armed == 1 {
+                                self.alarm.disable();
```
I did a little double-take here looking over this. It was really confusing see `num_armed` as non-zero leading to disabling the timer. Could this be re-written as something like:

```rust
let new_num_armed = self.num_armed.get() - 1;
self.num_armed.set(new_num_armed);
if new_num_armed == 0 {
```

This avoids needing to remember to add a `-1` everywhere that `num_armed` is referenced

alevy commented:
@phil-levis I don't know what the right names are basically.

Maybe the answer is that they should be in separate header files even though the implementation is the same?

It's called an "alarm" in userland because the system call interface it uses is really mostly similar to the AST controller on the SAM4L (which as we recall from previous efforst, is easier to virtualize than the timer interface). It exposes both cause, why not.

It used to be `timer.h` because it was originally `time.h`, which is a nice agnostic name, but that conflicts with libc's `time.h`, which is not great.

ppannuto commented on userland/libnrfserialization/serialization_tock.c:
```diff
@@ -150,6 +150,8 @@ void ble_serialization_callback (int callback_type, int rx_len, int c, void* oth
 // ser_app_hal_nrf51.c
 //
 
+#define STORM_INT 5
```
Where did this come from?

alevy replied:
Oh yeah... `STORM_INT` was being used but not defined anywhere. That's a dummy variable. I have no idea what it should be. Thanks for catching!

ppannuto commented on userland/libtock/alarm.c:
```diff
@@ -0,0 +1,220 @@
+    // Heap too small! Make it bigger
+    int new_capacity = (alarm_heap.capacity + 1) * 2;
+    alarm_heap.data = (alarm_t**)realloc(alarm_heap.data,
+        new_capacity * sizeof(alarm_t*));
```
Need to check that `realloc` succeeds here

ppannuto commented on userland/libtock/alarm.c:
```diff
@@ -0,0 +1,220 @@
+}
+
+alarm_t *alarm_at(uint32_t expiration, subscribe_cb cb, void* ud) {
+  alarm_t *alarm = (alarm_t*)malloc(sizeof(alarm_t));
```
Need to check return of `malloc` here

ppannuto commented on userland/libtock/alarm.c:
```diff
@@ -0,0 +1,220 @@
+alarm_repeating_t* alarm_every(uint32_t ms, subscribe_cb cb, void* ud) {
+  uint32_t interval = ms * alarm_internal_frequency() / 1000;
+
+  alarm_repeating_t* uud = (alarm_repeating_t*)malloc(sizeof(alarm_repeating_t));
```
and here

ppannuto replied:
also, what's `uud`? We could probably afford some more letters here :P

alevy replied:
No! Each letter in C costs a dollar!

ppannuto commented on userland/libtock/alarm.c:
```diff
@@ -0,0 +1,220 @@
+}
+
+alarm_repeating_t* alarm_every(uint32_t ms, subscribe_cb cb, void* ud) {
+  uint32_t interval = ms * alarm_internal_frequency() / 1000;
```
I _think_ that `*` and `/` have equal precedence in C, which will have these evaluate left to right, which could possibly result in unexpected overflows for large intervals. Consider `ms * (alarm_internal_frequency() / 1000)`

alevy replied:
Though actually the goal was to avoid truncating the division when the frequency is not a mutliple of 1000.

alevy commented on userland/libtock/alarm.h:
```diff
@@ -0,0 +1,112 @@
+ *
+ * \param alarm
+ */
+void alarm_cancel(alarm_t*);
```
well... sort of... technically the `repeating_alarm_t` structure has an internal reference to an underlying `alarm_t` that you can get at, which will work if you pass it in, but the client doesn't know the fields of the struct... Could just replicate the function or do something clever with the struct definitions to let pass in either kind of struct (C-style inheritance)

ppannuto commented on userland/libtock/alarm.c:
```diff
@@ -0,0 +1,220 @@
+void alarm_cancel(alarm_t* alarm) {
+  // Removing from a heap is tricky, so just remove the callback and let it get
+  // lazily removed.
+  alarm->callback = NULL;
```
I may be missing something here, but how does this work for repeating timers? It looks like line 129 is a no-op if an alarm triggers and callback is NULL, but won't that keep happening forever for a repeating timer?

alevy replied:
A repeating timer is basically a wrapper around a regular alarm where the callback for the alarm is a shared "repeating_cb" callback that reschedules a new alarm. If the unerlying alarm's callback is set to NULL, then "repeating_cb" won't be called for that timer and it will just fade into oblivion.

ppannuto commented on userland/libtock/internal/alarm.h:
```diff
@@ -0,0 +1,55 @@
+ */
+int alarm_internal_absolute(uint32_t tics);
+
+int alarm_internal_stop(void);
```
This one doesn't get a comment? ;)

alevy commented:
It builds!

alevy commented:
@bradjc fixed. It was a stupid bug (wasn't passing the right userdata back up to the repeating callback and tests didn't pick it up because I actually had two apps running.

alevy commented:
# Status update
I believe this is re-ready to merge.

First @adkinsjd and @bradjc objected to explicitly having a virtual_alarm module as well as an alarm module since it both made the client interface in most cases needlessly more complex and risked clients using both, which is dangerous (because the low-level implementation doesn't virtualize and will effectively cancel any outsanding virtual alarms).

I moved the low-level implmentation (previously `alarm.h`) into `internal/alarm.h` and renamed `virtual_alarm.h` to `alarm.h`.

Next, @ppannuto @bradjc and @phil-levis pointed out several flaws that I addressed:

  1. Naming the userland virtual alarms module just `alarm.h` is bad because it really encapsulates what we elsewhere refer to as both alarms (expire at a particular clock value) and timers (expire within a certain amount of time). I separated the namespace into an `alarm.h` and `timer.h` module (with `alarm_` and `timer_` prefixes, respectively), though the implementation is fairly entangled so I kept that together in `alarm_timer.c` to avoid exposing shared internal stuff like the heap data structure.

  2. Both @bradjc and @ppanuto asked for comments in the kernel system call driver's `command` implementation. I added those as well as reorganized the match statement cases to be in sequential order (as opposed to the other they were added in).

  3. There was no real way to cancel a repeating timer, since it is stored in a wrapper struct `timer_repeating` with an internal pointer to an underlying `alarm` that is not exposed to the client. Splitting out the namespaces actually helped. Now there is a `timer_cancel(timer)` that cancels and frees the timer.

ppannuto commented on userland/examples/nrf51_temp_sensor/main.c:
```diff
@@ -1,8 +1,7 @@
+#include "temperature.h"
 #include <stdio.h>
 #include <timer.h>
```
nit: why'd this move / ""'s there looks wrong?

alevy replied:
idk, uncrustify changed it or something somehow, but it should be ""---it's a local header file.
